transport-classes-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java

```diff
@@ -0,0 +1,1111 @@
+
+        directBuf.writeBytes(buf, buf.readerIndex(), readableBytes);
+        ReferenceCountUtil.safeRelease(holder);
+        return directBuf;
+    }
+
+    private static ByteBuf newDirectBuffer0(Object holder, ByteBuf buf, ByteBufAllocator alloc, int capacity) {
+        final ByteBuf directBuf = alloc.directBuffer(capacity);
+        directBuf.writeBytes(buf, buf.readerIndex(), capacity);
+        ReferenceCountUtil.safeRelease(holder);
+        return directBuf;
+    }
+
+    /**
+     * Cancel all outstanding reads
+     *
+     * @param registration          the {@link IOUringIoRegistration}.
+     * @param numOutstandingReads   the number of outstanding reads.
+     */
+    protected abstract void cancelOutstandingReads(IOUringIoRegistration registration, int numOutstandingReads);
+
+    /**
+     * Cancel all outstanding writes
+     *
+     * @param registration          the {@link IOUringIoRegistration}.
+     * @param numOutstandingWrites   the number of outstanding writes.
+     */
+    protected abstract void cancelOutstandingWrites(IOUringIoRegistration registration, int numOutstandingWrites);
+
+    @Override
+    protected void doDisconnect() throws Exception {
+    }
+
+    private void freeRemoteAddressMemory() {
+        if (remoteAddressMemory != null) {
+            Buffer.free(remoteAddressMemory);
+            remoteAddressMemory = null;
+        }
+    }
+
+    private void freeMsgHdrArray() {
+        if (msgHdrMemoryArray != null) {
+            msgHdrMemoryArray.release();
+            msgHdrMemoryArray = null;
+        }
+    }
+
+    @Override
+    protected void doClose() throws Exception {
+        active = false;
+
+        if (registration != null) {
+            if (socket.markClosed()) {
+                int fd = fd().intValue();
+                IOUringIoOps ops = IOUringIoOps.newClose(fd, 0, registration.id(), nextOpsId());
+                registration.submit(ops);
+            }
+        } else {
+            // This one was never registered just use a syscall to close.
+            socket.close();
+            freeRemoteAddressMemory();
+            freeMsgHdrArray();
+        }
+    }
+
+    @Override
+    protected void doBeginRead() {
+        if (inputClosedSeenErrorOnRead) {
+            // We did see an error while reading and so closed the input. Stop reading.
+            return;
+        }
+        if (readPending) {
+            // We already have a read pending.
+            return;
+        }
+        if (inReadComplete) {
+            // We are currently in the readComplete(...) callback which might issue more reads by itself. Just
+            // mark it as a pending read. If readComplete(...) will not issue more reads itself it will pick up
+            // the readPending flag, reset it and call doBeginReadNow().
+            readPending = true;
+            return;
+        }
+        doBeginReadNow();
+    }
+
+    private void doBeginReadNow() {
+        if (socket.isBlocking()) {
```

Comment:
yes for the server we still use non-blocking because of this:

```
        // We don't use a blocking fd for the server channel at the moment as
        // there is no support for IORING_CQE_F_SOCK_NONEMPTY and IORING_ACCEPT_DONTWAIT
        // at the moment. Once these land in the kernel we should check if we can use these and if so make
        // the fd blocking to get rid of POLLIN etc.
        // See:
        //
        //  - https://lore.kernel.org/netdev/20240509180627.204155-1-axboe@kernel.dk/
        //  - https://lore.kernel.org/io-uring/20240508142725.91273-1-axboe@kernel.dk/
```

Reply:
We can create an issue for it once this is merged. And yes I will look into providing buffers as a followup. It's an optimization. And yes fast-poll might waste a buffer in "idle mode". That said this is "only" 64k realistically.